### ISSUE TITLE
Meta  AWS::CloudFormation::Interface 

### ISSUE THREAD

**Issue by rackerroush** [shalupov/idea-cloudformation#50]
I am getting a 'cannot resolve symbol' message when I used the following inside the meta section of my Cloud Formation Templates.

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudformation-interface.html

First guess is it has some kind of name conflict when a parameter name is seen in more than one place, or maybe the fact that Parameter* is used as a property in the AWS::CloudFormation::Interface

Example:
```
    
  "Metadata": {
    "AWS::CloudFormation::Interface" : {
      "ParameterLabel" : {
        "VCIDR" : { "default" : "What should VPC CIDR Range be?" }
      },
      "ParameterGroup" : [
        {
          "Label" : { "default" : "Network Configuration" },
          "Parameters" : ["AZCount", "PubSub1", "PubSub2","PrivSub1", "PrivSub2" ]
        },
        {
          "Label" : { "default":"NAT EC2 Conf" },
          "Parameters" : [ "Tenancy", "KeyPair","InstanceType" ]
        }
      ]
    }
  }
```

**Comment by shalupov** [comment]
Could you attach full example?

**Comment by rackerroush** [comment]
Sorry it took me so long to get back to you on this.
Context: your latest versions which load under the 2016.1 CE Version of IntelliJ do not have a conflict.  I've not wanted to buy in on the subscription model so I have not upgraded from version 14 of the IDE.  I suppose that means it would be a backport issue from here on out, but it looks like something changed between 0.3.52 and 0.4.16+

NOTE:  I was trying to load in 0.4.5 by downloading it and doing a "Install plugin from disk" but when I attempted to install the zip file I got a "incompatible with this installation" warning.  (same results with 0.4.13)  (0.4.16 & 0.4.18 seem to install fine)  I must be missing something with the way versioning works.

Admission:  I'm contributing to my own angst in this case by not buying into the subscription model.  I have enjoyed using IntelliJ and it was not cheap for just my own needs but I felt like I would only update if I had a compelling reason to do so now and again.  Continuous billing & infrequent usage, on my part makes that seem unpalatable.  

Put together a better example and post here.

Thank you for all your time on this project.  It has made working with CloudFormation templates a better experience. =)

**Comment by shalupov** [comment]
Thank you for the feedback! 0.4.18 should work with IDEA CE 2016.1 only.
Let's say 0.4.x versions are for EAP builds in the past.
I do not plan to actively support 0.3.x branch, maybe one more update or so.

As for your original issue, the plugin does not handle "AWS::CloudFormation::Interface" at all, it's just a json for it. I'll try to reproduce your issue on 0.3.52 if you post an entire file which reproduces the problem.

**Comment by rackerroush** [comment]
It does seem to be tied to AWS::CloudFormation::Interface in same fashion.  I took some targeted screen shots for clarity which I will post first followed by the json data.  I did not create the following template, but it does attribute the behavior I am seeing.

If the AWS::CloudFormation::Interface section is a part of the cloudformation
<img width="1018" alt="screen shot 2016-03-18 at 12 45 45" src="https://cloud.githubusercontent.com/assets/4241134/13887213/830d408e-ed08-11e5-9a7a-2d6a72785d6d.png">

I get 'cannot resolve symbol' error messages like this in the conditions sections:
<img width="653" alt="screen shot 2016-03-18 at 12 46 27" src="https://cloud.githubusercontent.com/assets/4241134/13887274/e0897494-ed08-11e5-8ffd-5168d11a81c9.png">

Same issues inside the Resources section:
<img width="1009" alt="screen shot 2016-03-18 at 12 46 45" src="https://cloud.githubusercontent.com/assets/4241134/13887304/ff7b8680-ed08-11e5-91b3-41133792bf54.png">


However, if I remove the AWS::CloudFormation::Interface section from the template the  'cannot resolve symbol' problem goes away.
<img width="1007" alt="screen shot 2016-03-18 at 12 47 23" src="https://cloud.githubusercontent.com/assets/4241134/13887337/24b748b2-ed09-11e5-8c05-a16b1b48446d.png">

**Comment by rackerroush** [comment]
```
{
  "AWSTemplateFormatVersion" : "2010-09-09",
  "Description" : "AWS Memcached or Redis Cluster for ElastiCache",
  "Metadata": {
    "Version": "v0.1.0",
    "AWS::CloudFormation::Interface": {
      "ParameterGroups" : [
        {
          "Label" : { "default" : "VPC Configuration" },
          "Parameters" : ["Environment", "VPCId", "ElastiCacheSubnets","ASGSecurityGroup" ]
        },
        {
          "Label" : { "default" : "ElastiCache Basic Configuration" },
          "Parameters" : ["ElastiCacheEngineType", "InstanceClass", "NumberOfNodes", "CacheClusterPort"]
        },
        {
          "Label" : { "default" : "ElastiCache Advanced Configuration" },
          "Parameters" : ["FailoverEnabled", "InstanceClass", "NumberOfNodes", "CacheClusterPort", "EnableNotifications", "MinorVersionUpgrade"]
        },
        {
          "Label" : { "default" : "Redis Specific" },
          "Parameters" : ["ReplicationGroupDescription"]
        },
        {
          "Label" : { "default" : "Memcached Specific" },
          "Parameters" : ["ClusterName"]
        },
        {
          "Label" : { "default" : "Backup / Maintenance" },
          "Parameters" : ["SnapshotRetentionLimit", "PreferredSnapshotWindow", "PreferredMaintenanceWindow"]
        }
      ]
    }
  },

  "Parameters": {
    "VPCId": {
      "Description": "Existing Virtual Private Cloud to use with the elasticache cluster",
      "Type": "AWS::EC2::VPC::Id"
    },
    "InstanceClass": {
      "Description": "The compute and memory capacity of the nodes within the ElastiCache cluster",
      "Default": "cache.m3.medium",
      "Type": "String",
      "AllowedValues": [
        "cache.t2.micro", "cache.t2.small", "cache.t2.medium",
        "cache.m3.medium", "cache.m3.large", "cache.m3.xlarge", "cache.m3.xlarge", "cache.m3.2xlarge",
        "cache.r3.large", "cache.r3.xlarge", "cache.r3.2xlarge", "cache.r3.4xlarge", "cache.r3.8xlarge"
      ]
    },
    "NumberOfNodes": {
      "Description": "The number of cache nodes within the ElastiCache cluster",
      "Default": "1",
      "Type": "Number",
      "MinValue": "1",
      "MaxValue": "100"
    },
    "CacheClusterPort": {
      "Description": "The number of cache nodes within the ElastiCache cluster",
      "Default": "6379",
      "Type": "Number",
      "MinValue": "1024",
      "MaxValue": "65535"
    },
    "ElastiCacheEngineType": {
      "Description": "The name of the cache engine to be used for this cache cluster",
      "Default": "redis28",
      "Type": "String",
      "AllowedValues": [ "memcached14", "redis28" ]
    },
    "ElastiCacheSubnets": {
      "Description": "Subnets for use with this cache cluster",
      "Type": "List<AWS::EC2::Subnet::Id>"
    },
    "ASGSecurityGroup": {
      "Description": "Existing security group being used by ec2 instances within the application autoscale group",
      "Type": "CommaDelimitedList",
      "Default": ""
    },
    "FailoverEnabled": {
      "Description": "Specifies if enable failover",
      "Type": "String",
      "Default": "true",
      "AllowedValues": [ "true", "false" ]
    },
    "MinorVersionUpgrade": {
      "Description": "Specifies version upgrade is enabled",
      "Type": "String",
      "Default": "true",
      "AllowedValues": [ "true", "false" ]
    },
    "Environment": {
      "Description": "Application environment for which this Cache Cluster is being created. e.g. Development/Production",
      "Type": "String",
      "Default": "Development",
      "AllowedValues": ["Development", "Integration", "PreProduction", "Production", "Staging", "Test"]
    },
    "ReplicationGroupDescription": {
      "Description": "Description of Replication Group (redis only)",
      "Type": "String",
      "Default": "Elasticache"
    },
    "ClusterName": {
      "Description": "Name of Cluster",
      "Type": "String",
      "Default": "MyCluster"
    },
    "SnapshotRetentionLimit": {
      "Default": "7",
      "Description" : "The number of days for which automated backups are retained. Setting this parameter to a positive number enables backups. Setting this parameter to 0 disables automated backups.",
      "Type": "Number",
      "MinValue": "0",
      "MaxValue": "35",
      "ConstraintDescription" : "must be between 0 and 35"
    },
    "PreferredSnapshotWindow": {
      "Default": "Sun:03:00-Sun:04:00",
      "Description" : "The daily time range during which automated backups are created if automated backups are enabled",
      "Type": "String"
    },
    "PreferredMaintenanceWindow": {
      "Default": "Sun:05:00-Sun:07:00",
      "Description" : "The daily time range during which automated backups are created if automated backups are enabled",
      "Type": "String"
    },
    "EnableNotifications" : {
      "Type" : "String",
      "Default" : "False",
      "AllowedValues" : ["False", "True"],
      "Description" : "Specifies whether to notify if an autoscaling event ocurrs or not."
    }
  },
  "Conditions" : {
    "NotificationsEnabled" : {"Fn::Equals" : [{"Ref" : "EnableNotifications"}, "True"]},
    "Memcached" : {"Fn::Equals" : [{"Ref" : "ElastiCacheEngineType"}, "memcached14"]},
    "Redis" : {"Fn::Equals" : [{"Ref" : "ElastiCacheEngineType"}, "redis28"]}
  },
  "Mappings": {
    "ElastiCacheEngine" : {
      "memcached14" : {
        "ParameterGroupName": "default.memcached1.4",
        "Version": "1.4.24",
        "Name": "memcached"
      },
      "redis28" : {
        "ParameterGroupName": "default.redis2.8",
        "Version": "2.8.24",
        "Name": "redis"
      }
    }
  },
  "Resources": {
    "ElastiCacheRepGroup": {
      "Type": "AWS::ElastiCache::ReplicationGroup",
      "Condition": "Redis",
      "Properties": {
        "NotificationTopicArn" : {
          "Fn::If" :
          [
            "NotificationsEnabled",
            {
              "Fn::Join" : [":",["arn","aws","sns",{ "Ref" : "AWS::Region" },{ "Ref": "AWS::AccountId" },"rackspace-support"]]
            }
          ,
            { "Ref" : "AWS::NoValue" }
          ]
        },
        "AutomaticFailoverEnabled": { "Ref": "FailoverEnabled" },
        "AutoMinorVersionUpgrade": { "Ref": "MinorVersionUpgrade" },
        "CacheNodeType": { "Ref": "InstanceClass" },
        "CacheParameterGroupName": {"Fn::FindInMap": [ "ElastiCacheEngine", { "Ref": "ElastiCacheEngineType" }, "ParameterGroupName" ]},
        "CacheSubnetGroupName": {
          "Ref": "ElastiCacheSubnetGroup"
        },
        "Engine": {"Fn::FindInMap": [ "ElastiCacheEngine", { "Ref": "ElastiCacheEngineType" }, "Name" ]},
        "EngineVersion": {"Fn::FindInMap": [ "ElastiCacheEngine", { "Ref": "ElastiCacheEngineType" }, "Version" ]},
        "NumCacheClusters": { "Ref": "NumberOfNodes" },
        "Port": { "Ref": "CacheClusterPort" },
        "PreferredMaintenanceWindow": {"Ref": "PreferredMaintenanceWindow"},
        "ReplicationGroupDescription": {"Ref": "ReplicationGroupDescription"},
        "SecurityGroupIds": { "Ref": "ASGSecurityGroup" },
        "SnapshotRetentionLimit": {"Ref": "SnapshotRetentionLimit"},
        "SnapshotWindow": {"Ref": "PreferredSnapshotWindow"}
      }
    },
    "CacheCluster": {
      "Type": "AWS::ElastiCache::CacheCluster",
      "Condition": "Memcached",
      "Properties": {
        "CacheNodeType": { "Ref": "InstanceClass" },
        "Engine": {"Fn::FindInMap": [ "ElastiCacheEngine", { "Ref": "ElastiCacheEngineType" }, "Name" ]},
        "AZMode": "cross-az",
        "Port": { "Ref": "CacheClusterPort" },
        "EngineVersion": {"Fn::FindInMap": [ "ElastiCacheEngine", { "Ref": "ElastiCacheEngineType" }, "Version" ]},
        "PreferredMaintenanceWindow": {"Ref": "PreferredMaintenanceWindow"},
        "CacheSubnetGroupName": { "Ref": "ElastiCacheSubnetGroup" },
        "VpcSecurityGroupIds": { "Ref": "ASGSecurityGroup"},
        "NumCacheNodes": { "Ref": "NumberOfNodes" },
        "Tags": [
          { "Key": "ServiceProvider", "Value": "Rackspace" },
          { "Key": "Name", "Value": { "Ref": "ClusterName" } },
          { "Key": "Environment", "Value": { "Ref": "Environment" } }
        ]
      }
    },
    "ElastiCacheSubnetGroup": {
      "Type": "AWS::ElastiCache::SubnetGroup",
      "Properties": {
        "Description": "Subnets available for use with the cache cluster",
        "SubnetIds": { "Ref": "ElastiCacheSubnets"}
      }
    }
  },
  "Outputs": {
    "MemcachedConnection" : { "Value" : { "Fn::Join": [ ":", [ { "Fn::GetAtt": [ "CacheCluster", "ConfigurationEndpoint.Address" ] }, { "Fn::GetAtt": [ "CacheCluster", "ConfigurationEndpoint.Port" ] } ] ] }, "Condition" : "Memcached" },
    "RedisConnection" : { "Value" : { "Fn::Join": [ ":", [ { "Fn::GetAtt": [ "ElastiCacheRepGroup", "PrimaryEndPoint.Address" ] }, { "Fn::GetAtt": [ "ElastiCacheRepGroup", "PrimaryEndPoint.Port" ] } ] ] }, "Condition" : "Redis" }
  }
}
```

**Comment by shalupov** [comment]
Ok, that's an interesting one.

JsonObject.findProperty in IDEA 14-15 returns any children with that name down the tree. In IDEA 2016.1 JsonObject.findProperty returns only immediate children (which is correct, I think).

Because of it in this file all parameters are resolved via "Parameters" of "ParameterGroups" element instead of top-level section.

**Comment by shalupov** [comment]
Also I found some tests were not running :)

**Comment by shalupov** [comment]
Released 0.3.53

**Comment by rackerroush** [comment]
Just fantastic!  Thank you @shalupov
